### PR TITLE
[No Squash] Show dep errors in Select Mods modal

### LINKED ISSUE #9878: Cant tell difference between enable/disabled (built-in?) mods
## Problem
New to minetest.  Latest stable Debian 10.x build.  I installed 0.4.17.1 from Debian repo, but later pulled down 5.3.0-dev this GitHub repo.  Both installed (different directories).

Found some mods i wanted to try,  Added them to ~/.minetest/mods (_after running into problems installing under ~/.minetest/clientmods_).  Created a new world (mapgen v7 -- "minetest game"); wanted to enable mods, but noticed a UI problem.

Here's a screenshot selecting some new mods, except _footprints_.  I can tell the difference between selected and non-selected mods.
1_mods_default

After clicking **Enable All**, only the mods I downloaded are enabled.
2_mods_all_enabled

After clicking **Disable All**, the mods I downloaded are disabled.
3_mods_all_disabled

But these built-in mods ... are they always enabled?  Always disabled?  What if I wanted to play without a built-in mod?  Should I remove the mod from **$STATIC_SHAREDIR**/games/minetest_game/mods/ -- Nah, that's just crazy talk, right?

_Maybe I should move all the built-in mods to ~/.minetest/mods for finer control ..._  more crazy talk, or is it?

An additional reason why I post this is because prior to downloading those 7 mods, I created a world and ran around in it for about 10 minutes.  I didnt see beds, boats, butterflies, doors, animals (or mobs), etc.  Just terrain.  Lots and lots of terrain.

## Suggested Solution(s)

I think there should be a distinction in the UI between enabled built-in mods -vs- disabled built-in mods.  Failing that, a **note** indicating that built-in mods are always enabled/disabled and cannot be toggled.

Thanks.

### LINKED ISSUE #11712: Mod manager needs better dependency searching
## Problem
You can't search unsatisfied dependencies (which may be provided by modpacks only)
Modpacks and dependencies are merged into one list causing confusion

## Solutions
Create separate tab list for dependencies and modpacks

Allow user to find dependencies manually without need to install any mods first

## Alternatives

Maybe add some kind of troubleshooter or something like that to the game also a solution

## Additional context

1. Cannot find the dependencies required for a specific modpack
2. A mod is installed but dependencies are still unsatisfied

EDIT by SmallJoker: Grammar, re-word.

### PR DESCRIPTION
**Pull request by rubenwardy**
- Refactor ModConfiguration to be a reusable component rather than being inherited
- Use ModConfiguration in Select Mods dialog to show errors

Fixes #9878
Fixes #11712

Related #3330

Missing dependency:

image

Cycle:

image

Errors in mod packs and games

image

## To do

- [ ] Add/modify unit tests
- [x] Show errors on modpacks/games

## How to test

- Check the Select Mod behaviour when:
    - A dep isn't installed
    - A dep isn't enabled
    - There's a cycle
- Check that "The following mods could not be found:" still works

https://github.com/minetest/minetest/files/8646053/mods.zip

### CODE REVIEW

**Comment by rubenwardy** [comment]
Bonus screenshot, this is what happens if you set up mods like in VanessaE's optional dependency issue, where an optional dep stops Minetest from loading:

image

**Comment by rubenwardy** [comment]
Test [mods.zip](https://github.com/minetest/minetest/files/8646053/mods.zip)

For some reason it won't let me upload this to the PR description

File: src/server/mods.h [hunk 1]
```diff
@@ -42,6 +44,23 @@ class ServerModManager : public ModConfiguration
...
+		return configuration.isConsistent();
+	}
+
+	inline void printUnsatisfiedModsError() const {
+		return configuration.printUnsatisfiedModsError();
+	}
+
 	/**
 	 * Recursively gets all paths of mod folders that can contain media files.
 	 *
```
Comment (sfan5) [inline src/server/mods.h:66]: can't comment directly on the PNGs below but you didn't optimize them

File: src/content/ModConfiguration.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	for (u32 i = 0; i < m_unsatisfied_mods.size(); ++i) {
+		existing_mods[m_unsatisfied_mods[i].name] = i;
+	}
+
+	// Add new mods
+	for (int want_from_modpack = 1; want_from_modpack >= 0; --want_from_modpack) {
+		// First iteration:
+		// Add all the mods that come from modpacks
+		// Second iteration:
+		// Add all the mods that didn't come from modpacks
```
[the remark this answers is not in the document]
Comment (rubenwardy) [reply src/content/ModConfiguration.cpp:55]:
```
for (int want_from_modpack = 1; want_from_modpack >= 0; --want_from_modpack) {

// -- snip --


	    if (mod.part_of_modpack != (bool)want_from_modpack)
        	continue;
```

This code hasn't been modified

File: src/content/ModConfiguration.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+{
+	Settings conf;
+	std::unordered_map<std::string, std::string> load_mod_names;
+
+	conf.readConfigFile(settings_path.c_str());
+	std::vector<std::string> names = conf.getNames();
+	for (const std::string &name : names) {
+		const auto &value = conf.get(name);
+		if (name.compare(0, 9, "load_mod_") == 0 && value != "false" &&
+				value != "nil")
```
[the remark this answers is not in the document]
Comment (rubenwardy) [reply src/content/ModConfiguration.cpp:120]: Really we want `is_no`, as we want to accept any string that isn't a falsy value. Idk if nil will ever appear

Also, this code exists in master

File: src/content/ModConfiguration.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+			load_mod_names[name.substr(9)] = value;
+	}
+
+	std::vector<ModSpec> addon_mods;
+	std::unordered_map<std::string, std::vector<std::string>> candidates;
+
+	for (const auto &modPath : modPaths) {
```
[the remark this answers is not in the document]
Comment (rubenwardy) [reply src/content/ModConfiguration.cpp:127]:
```c++
/*
	 * Iterate through all installed mods
	 * 
	 * If explicitly enabled using a virtual_path, add to `addon_mods`.
	 * If enabled by load_mod_modname = true, add to `candidates` to be resolved later.
	 * 
	 * If not enabled, add `load_mod_modname = false` to world.mt
	 */
	 ```

**Comment by rubenwardy** [comment]
Updated. I've made the changes as two new commits, one for each commit, to make it easy to review the changes. These should be squashed into the two commits before merging

File: src/content/ModConfiguration.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfan5) [reply src/content/ModConfiguration.cpp]: I noticed that too afterwards but I'm still wondering.
Maybe use `!is_yes` then?

File: src/script/common/c_content.cpp [hunk 5]
```diff
@@ -2154,3 +2154,36 @@ void push_collision_move_result(lua_State *L, const collisionMoveResult &res)
...
+
+	lua_pushstring(L, spec.virtual_path.c_str());
+	lua_setfield(L, -2, "virtual_path");
+
+	lua_newtable(L);
+	int i = 1;
+	for (const auto &dep : spec.unsatisfied_depends) {
+		lua_pushnumber(L, i);
+		lua_pushstring(L, dep.c_str());
+		lua_settable(L, -3);
```
Comment (sfan5) [inline src/script/common/c_content.cpp:2186]: you could use `lua_rawseti` here
example code I quickly grepped for:
```cpp
1922-	lua_newtable(L);
1923-	s16 i = 1;
1924-	for (const std::string &texture : skybox_params.textures) {
1925-		lua_pushlstring(L, texture.c_str(), texture.size());
1926:		lua_rawseti(L, -2, i++);
1927-	}
```

File: src/script/lua_api/l_mainmenu.cpp [hunk 6]
```diff
@@ -409,6 +409,99 @@ int ModApiMainMenu::l_get_content_info(lua_State *L)
 	return 1;
 }
 
+/******************************************************************************/
+int ModApiMainMenu::l_check_mod_configuration(lua_State *L)
+{
+	std::string worldpath = luaL_checkstring(L, 1);
+
+	ModConfiguration modmgr{};
```
Comment (sfan5) [inline src/script/lua_api/l_mainmenu.cpp:417]:
```suggestion
	ModConfiguration modmgr;
```

File: src/script/lua_api/l_mainmenu.cpp [hunk 7]
```diff
@@ -409,6 +409,99 @@ int ModApiMainMenu::l_get_content_info(lua_State *L)
...
+
+	luaL_checktype(L, 2, LUA_TTABLE);
+
+	lua_pushnil(L);
+	while (lua_next(L, 2)) {
+		// Ignore non-string keys
+		if (lua_type(L, -2) != LUA_TSTRING) {
+			lua_pop(L, 1); // pop the value
+			continue;
+		}
```
Comment (sfan5) [inline src/script/lua_api/l_mainmenu.cpp:435]: better to `sanity_check` since this is an internal api?

**Comment by sfan5** [comment]
Bogus warnings:
<pre>
2022-07-10 11:55:41: WARNING[Main]: Mod name conflict detected: "default"
2022-07-10 11:55:41: WARNING[Main]: Will not load: /home/stefan/minetest/bin/../games/minetest_game/mods/default
2022-07-10 11:55:41: WARNING[Main]: Overridden by: /home/stefan/minetest/bin/../games/minetest_game/mods/default
</pre>

**Comment by rubenwardy** [comment]
> Bogus warnings:
> 
> 2022-07-10 11:55:41: WARNING[Main]: Mod name conflict detected: "default"
> 2022-07-10 11:55:41: WARNING[Main]: Will not load: /home/stefan/minetest/bin/../games/minetest_game/mods/default
> 2022-07-10 11:55:41: WARNING[Main]: Overridden by: /home/stefan/minetest/bin/../games/minetest_game/mods/default

I can't reproduce this. Are you using RUN_IN_PLACE=1?

**Comment by sfan5** [comment]
yes

**Comment by rubenwardy** [comment]
Is that when opening the select mods dialog or when loading a mod? Do you have any mods installed/enabled?

**Comment by sfan5** [comment]
Narrowed it down to toggling mesecons on a MTG world (in the dialog), no other mods installed.

**Comment by rubenwardy** [comment]
Fixed. The recursive dependency enabler was also setting `.enable = true` on game mods

**Review by sfan5 — APPROVED** [review]
Tested last time, should be fine now

**Pull request merged**